### PR TITLE
[RISCV] Improve vmsge and vmsgeu selection

### LINKED ISSUE #114505: [RISCV] vmsge_vx intrinsics with a large immediate should do vmsgt(x,imm-1)
The following two functions have identical behavior:

```c
#include<riscv_vector.h>
vbool32_t vmsgeu(vuint32m1_t op1, size_t vl) {
    return __riscv_vmsgeu_vx_u32m1_b32(op1, 100, vl);
}
vbool32_t vmsgtu(vuint32m1_t op1, size_t vl) {
    return __riscv_vmsgtu_vx_u32m1_b32(op1, 99, vl);
}
```
but clang generates different code for them, less efficient for the vmsgeu case:
```asm
vmsgeu:
        li      a1, 100
        vsetvli zero, a0, e32, m1, ta, ma
        vmsltu.vx       v8, v8, a1
        vmnot.m v0, v8
        ret

vmsgtu:
        li      a1, 99
        vsetvli zero, a0, e32, m1, ta, ma
        vmsgtu.vx       v0, v8, a1
        ret
```
https://riscv.godbolt.org/z/63rTTanMY

Additionally, for a dynamic value known to not be the minimum value for the given integer type, it would presumably be better to decrement it in the GPRs than doing the mask negation, e.g. https://riscv.godbolt.org/z/14qPd643P (esp. if the decrement can be hoisted out of a loop).

### PR DESCRIPTION
**Pull request by jacquesguan**
Select vmsge(u) vs, C to vmsgt(u) vs, C-1 if C is not in the imm range and not the minimum value.

Fix https://github.com/llvm/llvm-project/issues/114505.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 1]
```diff
@@ -1616,28 +1616,37 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
...
-        int64_t CVal = C->getSExtValue();
+        IsCmpConstant = true;
+        CVal = C->getSExtValue();
         if (CVal >= -15 && CVal <= 16) {
           if (!IsUnsigned || CVal != 0)
             break;
-          IsCmpUnsignedZero = true;
+          IsCmpMinimum = true;
+        } else if (!IsUnsigned &&
+                   CVal == APSInt::getMinValue(Src1VT.getScalarSizeInBits(),
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:1635]: You can use INT64_MIN here

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:1635]: Err nevermind it needs to be element sized.

Comment (jacquesguan) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp]: Addressed

**Review by topperc — APPROVED** [review]
LGTM

**Review by lukel97 — APPROVED** [review]

**Pull request merged**